api/v1_projects.inc

```diff
@@ -656,3 +657,85 @@ function _get_enabled_filter($query_params)
         return $bool;
     }
 }
+
+//---------------------------------------------------------------------------
+// Proofreading functions
+
+// checkout a page
+function api_v1_project_checkout($method, $data, array $query_params)
+{
+    $proof_project = new ProofProject($data[":projectid"], get_project_state($query_params));
+    return $proof_project->checkout();
+}
+
+function api_v1_project_validatetext($method, $data, array $query_params)
+{
+    $text_checker = new TextChecker($data[":projectid"]);
+    return $text_checker->analyse(receive_text());
+}
+
+function api_v1_project_page($method, $data, $query_params)
+{
+    $proof_project_page = new ProofProjectPage($data[":projectid"], get_project_state($query_params), $data[":pagename"], get_page_state($query_params));
+    if ($method == "GET") {
+        return $proof_project_page->render_page_text("current");
+    }
+    // must be PUT
+    $page_action = $query_params['pageaction'] ?? null;
+    if (null === $page_action) {
+        throw new BadRequest("No page action found in request.");
+    }
+    switch($page_action) {
+        case 'resume':
+            return $proof_project_page->resume_page();
+        case 'saveasdone':
+            return $proof_project_page->attempt_save_as_done(receive_text());
+        case 'savetemp':
+            return $proof_project_page->save_as_in_progress(receive_text());
+        case 'saverevert':
+            return $proof_project_page->save_and_revert(receive_text());
+        case 'return':
+            return $proof_project_page->return_to_round();
+        default:
+            throw new BadRequest("$page_action is not a valid page action.");
+            break;
+    }
+}
+
+function get_project_state($query_params)
+{
+    global $PROJECT_STATES_IN_ORDER;
+
+    $state = $query_params["state"] ?? null;
+    if (null === $state) {
+        throw new ProjectNoStateGivenException("No project state found in request.");
+    }
+    if (!in_array($state, $PROJECT_STATES_IN_ORDER)) {
+        throw new ProjectInvalidStateException(sprintf("%s is not a valid project state", $state));
+    }
+    return $state;
+}
+
+function get_page_state($query_params)
+{
+    global $PAGE_STATES_IN_ORDER;
+
+    $page_state = $query_params['pagestate'] ?? null;
+    if (null === $page_state) {
+        throw new ProjectNoPageStateGivenException("No page state found in request.");
+    }
+    if (!in_array($page_state, $PAGE_STATES_IN_ORDER)) {
```

Comment:
This also seems pointless.

Reply:
Done this.

Reply:
Done this. There are, I think, only 25 possible page states.
We could go further and validate against the actual page state. Consider that when we validate the page name we check it against the array of all the pages in the project. But we could instead search for it in the projectid table and create a Page object containing its name and state and user for the round since we now know the round. We could then validate also that the logged-in user owns the page. This parallels what we do for projects: rather than get a list of all projects and check if it is in the list, we search for it by id and create an object containing relevant data.